lib/src/zenohUTransport.cpp

```diff
@@ -22,190 +22,113 @@
-
-    return status;
+    spdlog::info("ZenohUTransport destructor done");
 }
 
-UStatus ZenohUTransport::send(const UUri &uri,
-                              const UPayload &payload,
-                              const UAttributes &attributes) noexcept {
-    UStatus status;
-
-    if (0 == refCount_) {
-        spdlog::error("ZenohUTransport is not initialized");
-        status.set_code(UCode::UNAVAILABLE);
-        return status;
-    }
-
-    if (true == termPending_) {
-        spdlog::error("ZenohUTransport is marked for termination");
-        status.set_code(UCode::UNAVAILABLE);
-        return status;
-    }
+UStatus ZenohUTransport::send(const UMessage &message) noexcept {
     
-    if ((0 == payload.size()) || (nullptr == payload.data())) {
-        spdlog::error("invalid paylooad");
-        status.set_code(UCode::UNAVAILABLE);
-        return status;
-    }
+    UStatus status;
 
-    if (false == isRPCMethod(uri.resource())) {
-        status.set_code(sendPublish(uri, payload, attributes));
+    if (UMessageType::UMESSAGE_TYPE_PUBLISH == message.attributes().type()) {
+        status.set_code(sendPublish(message));
+    } else if (UMessageType::UMESSAGE_TYPE_RESPONSE == message.attributes().type()) {
+         if (false == isRPCMethod(message.attributes().sink())) {
+            spdlog::error("message defined as response but the URI is not RPC ");
+            return status;
+         }
+        status.set_code(sendQueryable(message));
     } else {
-        status.set_code(sendQueryable(uri, payload, attributes));
+        spdlog::error("message type is not supported");
+        status.set_code(UCode::INTERNAL);
     }
 
     return status;
 }
 
-UCode ZenohUTransport::sendPublish(const UUri &uri, 
-                                   const UPayload &payload,
-                                   const UAttributes &attributes) noexcept {
-    UCode status;
+UCode ZenohUTransport::sendPublish(const UMessage &message) noexcept {
 
-    status = UCode::UNAVAILABLE;
+    UCode status = UCode::UNAVAILABLE;
 
+    pendingSendRefCnt_.fetch_add(1);
     do {
-
-        if (UMessageType::PUBLISH != attributes.type()) {
-
-            spdlog::error("Wrong message type = {}", UMessageTypeToString(attributes.type()).value());
-            return UCode::INVALID_ARGUMENT;
+      
+        if ((0 == message.payload().size()) || (nullptr == message.payload().data())) {
+            spdlog::error("payload not valid");
+            break;
         }
 
-        /* get hash and check if the publisher for the URI is already exists */
-        auto uriHash = std::hash<std::string>{}(LongUriSerializer::serialize(uri)); 
-        auto handleInfo = pubHandleMap_.find(uriHash); 
-
-        /* increment the number of pending send operations*/
-        pendingSendRefCnt_.fetch_add(1);
+        /* get key and check if the publisher for the URI is already exists */
+        auto key = uprotocol::uri::toZenohKeyString(message.attributes().source());
```

Comment:
you can move this code to the Send API , and pass the key to either SendPub or SendQuery API , just to avoid duplication

Reply:
The goal of this PR is to keep the change from long to micro URIs as small as possible. This is where long URIs were serialized before. I'd be happy to make a follow-up PR to restructure and clean up the code some.

Reply:
@gregmedd  please open an issue for tracking